gpkit/nomials/math.py

```diff
@@ -40,94 +33,36 @@ class Signomial(Nomial):
-            cs = [cs]
-            exps = [HashVector(exp)]  # pylint: disable=redefined-variable-type
-        elif isinstance(exps, Nomial):
-            simplify = False
-            cs = exps.cs  # pylint: disable=no-member
-            exps = exps.exps  # pylint: disable=no-member
-        else:
-            try:
-                # test for presence of length and identical lengths
-                assert len(cs) == len(exps)
-                exps_ = list(range(len(exps)))
-                if not all(isinstance(c, Quantity) for c in cs):
-                    try:
-                        cs = np.array(cs, dtype='float')
-                    except ValueError:
-                        raise ValueError("cannot add dimensioned and"
-                                         " dimensionless monomials together.")
-                else:
-                    units = Quantity(1, cs[0].units)
-                    if units.dimensionless:
-                        cs = [c * ureg.dimensionless for c in cs]
-                        units = ureg.dimensionless
-                    try:
-                        cs = [c.to(units).magnitude for c in cs] * units
-                    except DimensionalityError:
-                        raise ValueError("cannot add monomials of"
-                                         " different units together")
-                for i, k in enumerate(exps):
-                    exps_[i] = HashVector(k)
-                    for key in exps_[i]:
-                        if isinstance(key, Strings+(Monomial,)):
-                            exps_[i][VarKey(key)] = exps_[i].pop(key)
-                exps = tuple(exps_)
-            except AssertionError:
-                raise TypeError("cs and exps must have the same length.")
-
-        if isinstance(units, Quantity):
-            if not isinstance(cs, Quantity):
-                cs = cs*units
-            else:
-                cs = cs.to(units)
-
-        # init NomialData to create self.exps, self.cs, and so on
-        super(Signomial, self).__init__(exps, cs, simplify=simplify)
-
+    _c = _exp = None  # pylint: disable=invalid-name
+
+    def __init__(self, hmap=None, cs=1, require_positive=True, **descr):  # pylint: disable=too-many-statements,too-many-branches
+        if not isinstance(hmap, NomialMap):
+            if hasattr(hmap, "hmap"):
+                hmap = hmap.hmap
+            elif isinstance(hmap, Numbers):
+                hmap_ = NomialMap([(EMPTY_EXP, mag(hmap))])
+                hmap_.set_units(hmap)
+                hmap = hmap_
+            elif hmap is None:
+                hmap = VarKey(**descr).hmap
+            elif isinstance(hmap, Strings):
+                hmap = VarKey(hmap, **descr).hmap
+            elif isinstance(hmap, dict):
+                exp = HashVector({VarKey(k): v for k, v in hmap.items()})
+                hmap = NomialMap({exp: mag(cs)})
+                hmap.set_units(cs)
+                hmap.remove_zeros()
```

Comment:
Yeah, in the three-year-old unit tests...potentially an area for some deprecation